File: KalkJava/geometrykalk/Tetrahedron.java
package geometrykalk;

import java.util.List;


public class Tetrahedron extends PlatonicSolid {
	
	
	/**
	 * Builds a tetrahedron starting from its face
	 */
	protected void createShape(Plane p) throws KalkException {
		//creates eight points representing the vertices of the tetrahedron (in the first quadrant)
		if(p instanceof EquilateralTriangle) {
			EquilateralTriangle t = (EquilateralTriangle) p;
	        List<Point2> vP = t.getPoints();
	        for(Point2 point : vP) {
	            setPoint(point.getX(), point.getY(), 0);
	        }
	        Point2 c = t.centreOfSymmetry(); //centre of the base shape
	        double h = height(t.side()); //height of the tetrahedron
	        setPoint(c, h);
	    } else throw new KalkException("It is not possible to build the object from the current plane shape.");
	}

	
	/**
	 * Constructor with one double type parameter
	 * @param s represents the length of the edge
	 * @throws KalkException if r is not a positive value
	 */
	public Tetrahedron(double e) throws KalkException {
		super("Tetrahedron");
		EquilateralTriangle s = new EquilateralTriangle(e);
		createShape(s);
	}
	
	
	/**
	 * Default constructor
	 */
	public Tetrahedron() throws KalkException {
		this(1);
	}
	
	
	/**
	 * Constructor with one parameter, which represents its face.
	 * Used for 2D->3D conversion
	 */
	public Tetrahedron(EquilateralTriangle t) throws KalkException {
		super("Tetrahedron");
		createShape(t);
	}

	
	/**
	 * @return the height of the tetrahedron based on its edge @param e
	 */
	public static double height(double e) { 
	    return Math.sqrt(6) / 3 * e;
	}

	
	/**
	 * @return the surface of the solid, excluding the area of its base and top faces
	 */
	public double lateralSurface() {
	    return (3 * Math.sqrt(3)) / 4 * Math.pow(edge(), 2);
	}
	
	
	public int schlafliSymbol_q() {
		return 3;
	}
	
	
	public double inRadius() throws KalkException {
		return circumRadius() / 3;
	}
	
	
	public double circumRadius() {
	    return edge() * Math.sqrt(6) / 4;
	}
	
	
	public double volume() {
	    return Math.sqrt(2) / 12 * Math.pow(edge(), 3);

	}
	

	public EquilateralTriangle face() throws KalkException {
	    return new EquilateralTriangle(edge());
	}

	
	public Point3 centreOfSymmetry() throws KalkException {
		List<Point3> lP = getPoints();
		double z = 0.0;
		for(Point3 point : lP) {
	        z += point.getZ();
	    }
	    Point2 p = face().centreOfSymmetry();
	    return new Point3(p.getX(), p.getY(), z / 4);
	}

	
	public void expand(double d) throws KalkException {
		if(d < 0)
			throw new KalkException("Negative values not allowed.");
		Tetrahedron t = new Tetrahedron(edge() + edge() * d / 100);
	    this.setEqual(t);
	}
	
	
	public void reduce(double d) throws KalkException {
		if(d < 0 || d >= 100)
			throw new KalkException("Value out of range. It must be between 0 and 100.");
		Tetrahedron t = new Tetrahedron(edge() + edge() * d / 100);
	    this.equals(t);
	}
	

	public Tetrahedron sum(Solid s) throws KalkException {
		if(s.getClass() != getClass())
			throw new KalkException("The two shapes are not of the same type.");
		Tetrahedron t = (Tetrahedron) s;
	    double newEdge = edge() + t.edge();
	    return new Tetrahedron(newEdge);
	}
	
	
	public Tetrahedron difference(Solid s) throws KalkException {
		if(s.getClass() != getClass())
			throw new KalkException("The two shapes are not of the same type.");
		Tetrahedron t = (Tetrahedron) s;
	    EquilateralTriangle faceDifference = face().difference(t.face()); //it may throw exceptions
	    return new Tetrahedron(faceDifference);
	}
	
	
	public Tetrahedron multiplication(Solid s) throws KalkException {
		if(s.getClass() != getClass())
			throw new KalkException("The two shapes are not of the same type.");
		Tetrahedron c = (Tetrahedron) s;
	    double newEdge = edge() * c.edge();
	    return new Tetrahedron(newEdge);
	}
	
	
	public Tetrahedron division(Solid s) throws KalkException {
		if(s.getClass() != getClass())
			throw new KalkException("The two shapes are not of the same type.");
		Tetrahedron c = (Tetrahedron) s;

	    //both operands are greater than 0
	    double newEdge = edge() / c.edge();
	    return new Tetrahedron(newEdge);
	}

}

File: Figure/Model/tetrahedron.cpp
#include "tetrahedron.h"
#include <math.h>
#include <typeinfo>


Tetrahedron::Tetrahedron(double e) : PlatonicSolid("Tetrahedron") {
    EquilateralTriangle* pT = new EquilateralTriangle(e); //it may throw an exception
    createShape(pT);
    delete pT;
}


Tetrahedron::Tetrahedron(EquilateralTriangle* pT) : PlatonicSolid("Tetrahedron") {
    createShape(pT);
}


void Tetrahedron::createShape(const Plane* s) {
    //creates five points representing the vertices of the tetrahedron (in the first quadrant)
    const EquilateralTriangle* pT = dynamic_cast<const EquilateralTriangle*>(s);
    if(pT) {
        const std::vector<Point2> vP = pT->getPoints();
        for(std::vector<Point2>::const_iterator it = vP.begin(); it != vP.end(); ++it) {
            setPoint(it->getX(), it->getY(), 0);
        }
        Point2* c = pT->centreOfSymmetry(); //centre of the base shape
        double h = height(pT->side()); //height of the tetrahedron
        setPoint(*c, h);
    } else throw WrongShape();
}


double Tetrahedron::height(double edge) {
    return sqrt(6) / 3 * edge;
}


double Tetrahedron::lateralSurface() const {
    return (3 * sqrt(3)) / 4 * pow(edge(), 2);
}


int Tetrahedron::schlafliSymbol_q() const { return 3; }


double Tetrahedron::inRadius() const {
    return circumRadius() / 3;
}


double Tetrahedron::circumRadius() const {
    return edge() * sqrt(6) / 4;
}


double Tetrahedron::volume() const {
    return sqrt(2) / 12 * pow(edge(), 3);
}


EquilateralTriangle* Tetrahedron::face() const {
    return new EquilateralTriangle(edge());
}


Point3* Tetrahedron::centreOfSymmetry() const {
    const std::vector<Point3> vP = this->getPoints();
    double z = 0.0;
    for(std::vector<Point3>::const_iterator it = vP.begin(); it != vP.end(); ++it) {
        z += it->getZ();
    }
    Point2* pP = face()->centreOfSymmetry();
    return new Point3(pP->getX(), pP->getY(), z / 4);
}


void Tetrahedron::expand(double n) {
    if(n < 0) throw NegativeValue();
    Tetrahedron* pT = new Tetrahedron(edge() + edge() * n / 100);
    *this = *pT;
    delete pT;
}


void Tetrahedron::reduce(double n) {
    if(n < 0 || n >= 100) throw ValueOutOfRange();
    Tetrahedron* pT = new Tetrahedron(edge() - edge() * n / 100);
    *this = *pT;
    delete pT;
}


Tetrahedron* Tetrahedron::operator+(const Solid& s) const {
    if(typeid(s) != typeid(*this)) throw DifferentShapes();
    const Tetrahedron& rT = static_cast<const Tetrahedron&>(s);
    double newEdge = edge() + rT.edge();
    return new Tetrahedron(newEdge);
}


Tetrahedron* Tetrahedron::operator-(const Solid& s) const {
    if(typeid(s) != typeid(*this)) throw DifferentShapes();
    const Tetrahedron& rT = static_cast<const Tetrahedron&>(s);
    EquilateralTriangle* pT = *face() - *(rT.face()); //it may throw exceptions
    return new Tetrahedron(pT);
}


Tetrahedron* Tetrahedron::operator*(const Solid& s) const {
    if(typeid(s) != typeid(*this)) throw DifferentShapes();
    const Tetrahedron& rT = static_cast<const Tetrahedron&>(s);
    double newEdge = edge() * rT.edge();
    return new Tetrahedron(newEdge);
}


Tetrahedron* Tetrahedron::operator/(const Solid& s) const {
    if(typeid(s) != typeid(*this)) throw DifferentShapes();
    const Tetrahedron& rT = static_cast<const Tetrahedron&>(s);

    //both operands are greater than 0
    double newEdge = edge() / rT.edge();
    return new Tetrahedron(newEdge);
}
